pynicotine/plugins/leech_detector/__init__.py

```diff
@@ -71,72 +76,85 @@ def loaded_notification(self):
         if self.settings["num_folders"] < min_num_folders:
             self.settings["num_folders"] = min_num_folders
 
-        if self.settings["message"]:
-            self.str_action = "message leecher"
-        else:
-            self.str_action = "log leecher"
+        # Separate leechers set for faster membership checks
+        self.detected_leechers = set(self.settings["detected_leechers"])
 
         self.log(
-            "Ready to %ss, require users have a minimum of %d files in %d shared public folders.",
-            (self.str_action, self.settings["num_files"], self.settings["num_folders"])
+            "Require users have a minimum of %d files in %d shared public folders.",
+            (self.settings["num_files"], self.settings["num_folders"])
         )
 
     def upload_queued_notification(self, user, virtual_path, real_path):
 
-        if user in self.probed:
+        if user in self.probed_users:
             # We already have stats for this user.
             return
 
-        self.probed[user] = "requesting"
+        self.probed_users[user] = "requesting"
         self.core.request_user_stats(user)
+
         self.log("Getting statistics from the server for new user %s…", user)
 
     def user_stats_notification(self, user, stats):
 
-        if user not in self.probed:
+        if user not in self.probed_users:
             # We did not trigger this notification
             return
 
-        if self.probed[user] != "requesting":
+        num_files = stats["files"]
+        num_folders = stats["dirs"]
+        is_user_accepted = (num_files >= self.settings["num_files"] and num_folders >= self.settings["num_folders"])
+
+        if is_user_accepted and user in self.detected_leechers:
+            self.detected_leechers.remove(user)
+            self.settings["detected_leechers"].remove(user)
```

Comment:
`remove` -> `discard`

Reply:
Not while Nicotine+ is running, but I'll change to discard regardless.

Reply:
Maybe I need more sleep, list doesn't have discard()...